### PR TITLE
Lock OrderedDict loop aswell to avoid mutation errors. #694

### LINKED ISSUE #694: RuntimeError when mutating while iterating an OrderedDict
Hi all. I have a **multithreaded** test which is yielding me some `RuntimeErrors (RuntimeError: OrderedDict mutated during iteration)`. I couldn't isolate the issue yet, but I thought it would be helpful to already share it. I found on `store.py` there's a lock being used for the `setitem` but not for the `delitem`. That ringed me a bell but it didn't fix the issue. 

**Version: 3.22.0**

```Python
Traceback (most recent call last):
  File "[...]", line 278, in safe_bulk_write
    result = [...].bulk_write(ops)
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 1794, in bulk_write
    return BulkWriteResult(bulk.execute(), True)
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 322, in execute
    op_result = execute_func()
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 169, in exec_update
    result = collection._update(spec=selector, document=document,
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 621, in _update
    for existing_document in itertools.chain(self._iter_documents(spec), [None]):
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 1263, in _iter_documents
    return (document for document in list(self._store.documents)
  File "[...]lib/python3.8/site-packages/mongomock/store.py", line 128, in documents
    for doc in six.itervalues(self._documents):
RuntimeError: OrderedDict mutated during iteration

Traceback (most recent call last):
  File "[...]", line 278, in safe_bulk_write
    result = [...].bulk_write(ops)
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 1794, in bulk_write
    return BulkWriteResult(bulk.execute(), True)
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 322, in execute
    op_result = execute_func()
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 169, in exec_update
    result = collection._update(spec=selector, document=document,
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 621, in _update
    for existing_document in itertools.chain(self._iter_documents(spec), [None]):
  File "[...]lib/python3.8/site-packages/mongomock/collection.py", line 1263, in _iter_documents
    return (document for document in list(self._store.documents)
  File "[...]lib/python3.8/site-packages/mongomock/store.py", line 127, in documents
    self._remove_expired_documents()
  File "[...]lib/python3.8/site-packages/mongomock/store.py", line 133, in _remove_expired_documents
    self._expire_documents(index)
  File "[...]lib/python3.8/site-packages/mongomock/store.py", line 152, in _expire_documents
    expired_ids = [
  File "[...]lib/python3.8/site-packages/mongomock/store.py", line 152, in <listcomp>
    expired_ids = [
RuntimeError: OrderedDict mutated during iteration
```

It's hard to provide some test data as it's not an error that happens often and the testcase is a bit noisy, but I'll let you know if I manage to isolate it.

### PR DESCRIPTION
**Pull request by denivyruck**
Had a threaded test in order to catch the issue but couldn't come up with an unit test for it.

At least fixes #694 for now.

### CODE REVIEW

File: mongomock/aggregate.py [hunk 1]
```diff
@@ -1308,7 +1308,7 @@ def _handle_project_stage(in_collection, unused_database, options):
             except KeyError:
                 # Ignore missing key.
                 pass
-    if (method == 'include') == (include_id is not False and include_id is not 0):
+    if (method == 'include') == (include_id and include_id != 0):
```
Comment (pcorpet) [inline mongomock/aggregate.py]: I'm not sure this change is related. Please revert and send in a different PR if you meant it

**Review by pcorpet — CHANGES_REQUESTED** [review]

File: mongomock/store.py [hunk 2]
```diff
@@ -149,10 +149,13 @@ def _expire_documents(self, index):
         # "key" structure = list of (field name, direction) tuples
         ttl_field_name = index['key'][0][0]
         ttl_now = mongomock.utcnow()
-        expired_ids = [
-            doc['_id'] for doc in six.itervalues(self._documents)
-            if self._value_meets_expiry(doc.get(ttl_field_name), expiry, ttl_now)
-        ]
+
+        # 694 - Lock the whole iteration to avoid RuntimeError when mutating OrderedDicts
+        with lock:
```
Comment (pcorpet) [inline mongomock/store.py]: You would also need to lock anywhere iterating over documents, especially the @property document above.

I think a multiple read lock approach would be the right fix instead.
https://code.activestate.com/recipes/577803-reader-writer-lock-with-priority-for-writers/

Comment (denivyruck) [reply mongomock/store.py]: Thanks @pcorpet. I'll try implementing it soon! :)

File: mongomock/aggregate.py (hunk 1 above)
Comment (denivyruck) [reply mongomock/aggregate.py]: Ops, was a warning I was receiving. Will revert

Comment (pcorpet) [reply mongomock/aggregate.py]: I agree that's annoying. I tried to silence the warning in 997adc6

File: mongomock/store.py (hunk 2 above)
Comment (denivyruck) [reply mongomock/store.py]: @pcorpet I've added the plain implementation with just some context managers as helpers. Tried to do a wrap around OrderedDict but it was 2x slower it seems.

**Comment by denivyruck** [comment]
There seems to be some deadlocks, I'm investigating.

**Comment by denivyruck** [comment]
Added a small test for the lock bug. I'm not familiar with the project structure so let me know if you want the code/tests relocated 👍

File: mongomock/store.py [hunk 3]
```diff
@@ -1,11 +1,91 @@
...
+from contextlib import contextmanager
+
 import mongomock  # Used for utcnow - please see https://github.com/mongomock/mongomock#utcnow
 import six
 import six.moves
 import threading
 
-lock = threading.RLock()
+
+class RWLock:
```
Comment (pcorpet) [inline mongomock/store.py]: Please move the thread specific classes to a mongomock/thread module

File: mongomock/store.py [hunk 4]
```diff
@@ -1,11 +1,91 @@
...
+    """Lock for enabling multiple readers but only 1 exclusive writer
+    to access a resource
+
+    Source: https://cutt.ly/Ij70qaq
+    """
+
+    def __init__(self):
+        self._read_switch = _LightSwitch()
+        self._write_switch = _LightSwitch()
+        self._no_readers = threading.Lock()
```
Comment (pcorpet) [inline mongomock/store.py]: I suggest you use RLock instead Lock so that a thread can actually use it multiple times.

File: tests/test_store.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from mongomock.store import RWLock
+
+
+def test_rwlock_exception():
```
Comment (pcorpet) [inline tests/test_store.py]: Please use a `unittest.TestCase` and `self.assertFalse` instead of `assert not`

**Review by pcorpet — CHANGES_REQUESTED** [review]
Thanks for adding the code

File: mongomock/store.py (hunk 4 above)
Comment (denivyruck) [reply mongomock/store.py]: It seems to be already covered. For instance, it's possible to do:
```
lock = RWLock()
with lock.reader():
    with lock.reader():
        ...
```
And it doesn't block the thread. Is that what you meant? 

I now added the `RLock` to the `_readers_queue`. Other than that it changes the behavior (exceptions, deadlocks, broken tests and panic lol).

File: mongomock/store.py (hunk 3 above)
Comment (denivyruck) [reply mongomock/store.py]: Done

File: mongomock/store.py (hunk 4 above)
Comment (denivyruck) [reply mongomock/store.py]: This implementation from [wikipedia](https://en.wikipedia.org/wiki/Readers%E2%80%93writers_problem#Third_readers%E2%80%93writers_problem) might be cleaner (and more fair):
```python
from contextlib import contextmanager

import threading


class RWLock:

    def __init__(self):
        self.readcount = 0
        # Controls the access to readcount
        self.readcount_lock = threading.Lock()

        # Controls the access to the critical region
        self.resource = threading.Lock()

        # Queue for readers/writers
        self.servicequeue = threading.RLock()

    @contextmanager
    def reader(self):
        self._acquire_for_reader()
        try:
            yield
        except Exception:
            raise
        finally:
            self._release_for_reader()

    def _acquire_for_reader(self):
        # Get in line
        self.servicequeue.acquire()
        # Lock for updating active readers
        self.readcount_lock.acquire()

        self.readcount += 1
        if self.readcount == 1:
            # First reader locks the resource for other writers
            self.resource.acquire()

        # Allow next reader/writer
        self.servicequeue.release()
        # Release access to readcount
        self.readcount_lock.release()

    def _release_for_reader(self):
        # Get access to readcount
        self.readcount_lock.acquire()
        self.readcount -= 1
        if self.readcount == 0:
            # If no one else is reading, release the resource
            self.resource.release()
        # Release access to readcount
        self.readcount_lock.release()

    @contextmanager
    def writer(self):
        self._acquire_for_writer()
        try:
            yield
        except Exception:
            raise
        finally:
            self._release_for_writer()

    def _acquire_for_writer(self):
        # Get in line
        self.servicequeue.acquire()
        # Request resource
        self.resource.acquire()
        # Let next in line to be serviced
        self.servicequeue.release()

    def _release_for_writer(self):
        # Release for next reader/writer
        self.resource.release()
```

**Comment by Bogay** [comment]
@pcorpet @denivyruck  
Any progress on this? or anything I can help to get this PR merged?  
I really need this patch to help passing tests in my project.

**Comment by pcorpet** [comment]
This seems clean enough for me, @denivyruck can you rebase and I'll merge? Otherwise I'll do it next week.

**Comment by denivyruck** [comment]
@pcorpet done 👍

**Comment by pcorpet** [comment]
Please check pylint and pep8 on Travis and fix it. Also the pypy fail is scary: it probably got caught in a deadlock :-(

**Comment by denivyruck** [comment]
> Please check pylint and pep8 on Travis and fix it. Also the pypy fail is scary: it probably got caught in a deadlock :-(

gonna investigate this pypy issue now

**Comment by denivyruck** [comment]
@pcorpet done! Linting was getting the best of me
